proposals/NNNN-resource-attributes-in-dxil-and-spirv.md

```diff
@@ -0,0 +1,386 @@
+| TextureCUBE                     | Cube   | Unknown (2) | false   | false | r/o (1) | r/o    |
+| TextureCUBEArray                | Cube   | Unknown (2) | true    | false | r/o (1) | r/o    |
+| RWTexture1D                     | 1D     | Unknown (2) | false   | false | r/w (2) | r/w    |
+| RWTexture1DArray                | 1D     | Unknown (2) | true    | false | r/w (2) | r/w    |
+| RWTexture2D                     | 2D     | Unknown (2) | false   | false | r/w (2) | r/w    |
+| RWTexture2DArray                | 2D     | Unknown (2) | true    | false | r/w (2) | r/w    |
+| RWTexture2DMS                   | 2D     | Unknown (2) | false   | true  | r/w (2) | r/w    |
+| RWTexture2DMSArray              | 2D     | Unknown (2) | true    | true  | r/w (2) | r/w    |
+| RWTexture3D                     | 3D     | Unknown (2) | false   | false | r/w (2) | r/w    |
+| Buffer                          | Buffer | Unknown (2) | false   | false | r/o (1) | r/o    |
+| RWBuffer                        | Buffer | Unknown (2) | false   | false | r/w (2) | r/w    |
+
+DXC lowers [Constant/Texture/Structured/Byte Buffers] to [OpTypeStruct] in
+SPIR-V with a uniform storage class and various layout decorations. DXC tracks
+the necessary information to lower these in side structures, and we'll likely
+need to keep track of that information in new SPIR-V target extension types in
+our implementation.
+
+[Constant/Texture/Structured/Byte Buffers]:
+    https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/SPIR-V.rst#constant-texture-structured-byte-buffers
+[OpTypeStruct]:
+    https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpTypeStruct
+
+|                         | Vulkan Buffer Type | SPIR-V Decoration |
+|-------------------------|--------------------|-------------------|
+| ByteAddressBuffer       | Storage            | BufferBlock       |
+| RWByteAddressBuffer     | Storage            | BufferBlock       |
+| StructuredBuffer        | Storage            | BufferBlock       |
+| RWStructuredBuffer      | Storage            | BufferBlock       |
+| AppendStructuredBuffer  | Storage            | BufferBlock       |
+| ConsumeStructuredBuffer | Storage            | BufferBlock       |
+| cbuffer                 | Uniform            | Block             |
+| ConstantBuffer          | Uniform            | Block             |
+| tbuffer                 | Storage            | BufferBlock       |
+| TextureBuffer           | Storage            | BufferBlock       |
+
+The "RasterizerOrdered" (or ROV) resources in HLSL don't map directly to
+SPIR-V. In DXC, we emulate ROV by treating these the same as their non-ROV
+counterparts but wrapping accesses to the resource in critical sections using
+[SPV_EXT_fragment_shader_interlock]. We will probably need to extend the SPIR-V
+target extension types to represent these unless we want to emit the critical
+sections awkwardly early in the compiler.
```

Comment:
This is even harder than just wrapping the accesses in a critical section.

> If the entry point has any of the interlock ordering execution modes, it must dynamically execute each of OpBeginInvocationInterlockEXT and OpEndInvocationInterlockEXT, in that program order, exactly once.

So DXC generates a different begin and end instructions around each of the accesses, then a spirv-opt pass finds the small block of code that will mean the requirements above.

At first glance, I agree that we should somehow add to the type that it is an ROV. However, we will need to look in deatail a how we choose between the different executions mode, but all of that feels like a backend thing to me.